bench/bm_fft.cpp

```diff
@@ -75,45 +76,28 @@ testFFT() {
         gr::blocks::fft::FFT<T, DataSet<PrecisionType>, FFTw<FFTInDataType<T, PrecisionType>>> fft1({ { "fftSize", N } });
         std::ignore                                                                    = fft1.settings().apply_staged_parameters();
 
-        ::benchmark::benchmark<nRepetitions>(fmt::format("{} - fftw", type_name<T>())) = [&fft1, &signal] {
-            fft1.prepareInput(signal);
-            fft1.computeFFT();
+        std::vector<DataSet<PrecisionType>> resultingDataSets(1);
+        ::benchmark::benchmark<nRepetitions>(fmt::format("{} - fftw", type_name<T>())) = [&fft1, &signal, &resultingDataSets] {
+            expect(fair::graph::work_return_status_t::OK == fft1.process_bulk(signal, resultingDataSets));
         };
     }
     {
         gr::blocks::fft::FFT<T, DataSet<PrecisionType>, FFT<FFTInDataType<T, PrecisionType>>> fft1({ { "fftSize", N } });
         std::ignore                                                                   = fft1.settings().apply_staged_parameters();
-        ::benchmark::benchmark<nRepetitions>(fmt::format("{} - fft", type_name<T>())) = [&fft1, &signal] {
-            fft1.prepareInput(signal);
-            fft1.computeFFT();
+
+        std::vector<DataSet<PrecisionType>> resultingDataSets(1);
+        ::benchmark::benchmark<nRepetitions>(fmt::format("{} - fft", type_name<T>())) = [&fft1, &signal, &resultingDataSets] {
+            expect(fair::graph::work_return_status_t::OK == fft1.process_bulk(signal, resultingDataSets));
         };
     }
 
-    if constexpr (gr::blocks::fft::ComplexType<T>) {
+    if constexpr (gr::algorithm::ComplexType<T>) {
         ::benchmark::benchmark<nRepetitions>(fmt::format("{} - fftCT", type_name<T>())) = [&signal] {
             auto signalCopy = signal;
             computeFFFTCooleyTukey<T>(signalCopy);
         };
     }
 
-    {
-        gr::blocks::fft::FFT<T, DataSet<PrecisionType>, gr::algorithm::FFTw<FFTInDataType<T, PrecisionType>>> fft1({ { "fftSize", N } });
-        std::ignore = fft1.settings().apply_staged_parameters();
-        fft1.prepareInput(signal);
-        fft1.computeFFT();
-
-        ::benchmark::benchmark<nRepetitions>(fmt::format("{} - magnitude", type_name<T>())) = [&fft1] { fft1.computeMagnitudeSpectrum(); };
-        ::benchmark::benchmark<nRepetitions>(fmt::format("{} - phase", type_name<T>()))     = [&fft1] { fft1.computePhaseSpectrum(); };
-
-        double sum{ 0. };
-        fft1.computeMagnitudeSpectrum();
-        fft1.computePhaseSpectrum();
-        ::benchmark::benchmark<nRepetitions>(fmt::format("{} - dataset", type_name<T>())) = [&fft1, &sum] {
-            const auto ds1 = fft1.createDataset();
-            sum += static_cast<double>(ds1.signal_values[0]);
-        };
-    }
-
```

Comment:
Why don't you want to keep this performance measurements? At least as a reminder that calculation of phase and magnitude can be also at the same order as FFT itself?

Reply:
From a clean and lean code perspective, my decision was based on the following considerations:

1. The tests don't provide substantial additional value to the ones above, particularly as we're benchmarking our FFT against FFTW. Nuances such as magnitude and phase calculations might be better placed in a dedicated test suite.
2. The granular breakdown, often into one-liner methods, makes the FFT GR block's core functionality harder to read for new developers or those unfamiliar with the intricacies of FFT. Or expressed differently, since this is supposed to be a reference implementation/benchmark, what would the library become for hundreds of more blocks that we need to maintain if we maintain this high verbosity?
3. The current tests don't compare various implementations.
4. If the phase and magnitude calculations are potential performance bottlenecks, they should be benchmarked as stand-alone functions.
5. In striving for clean and lean code, it's vital that our benchmarks reflect the primary use cases and core functionalities of the components we're testing. Introducing benchmarks for subsidiary operations could detract from this primary focus.

I just wanted to let you know that your existing tests are good. At the same time, our tests must be clear, targeted, and underscore primary functionalities. Sometimes, *less is more*.

Reply:
Thanks a lot for a very detailed answer!